src/Microsoft.TemplateEngine.Utils/SemanticVersion.cs

```diff
@@ -0,0 +1,674 @@
+                return false;
+            }
+
+            version = new SemanticVersion(source, major, minor, patch, prerelease, metadata);
+            return true;
+        }
+
+        private static bool TryParseSegment(string source, ref int nextDot, out int value)
+        {
+            nextDot = source.IndexOf('.', nextDot);
+
+            if (nextDot < 0 || nextDot == source.Length - 1)
+            {
+                value = 0;
+                return false;
+            }
+
+            string segment = source.Substring(0, nextDot);
+
+            if (string.IsNullOrWhiteSpace(segment)
+                || (segment.Length > 1 && segment[0] == '0')
+                || !int.TryParse(segment, NumberStyles.None, CultureInfo.InvariantCulture, out value)
+                || value < 0)
+            {
+                value = 0;
+                return false;
+            }
+
+            return true;
+        }
+
+        private static bool ValidateAndExtractBuildMetadataSection(string source, int tail, out string metadata)
+        {
+            if (tail < source.Length && source[tail] == '+')
+            {
+                if (tail == source.Length - 1 || !IsValidMetadataSection(source, tail + 1, source.Length))
+                {
+                    metadata = null;
+                    return false;
+                }
+
+                metadata = source.Substring(tail + 1);
+                return true;
+            }
+
+            metadata = null;
+            return true;
+        }
+
+        private static bool ValidateAndExtractPrereleaseSection(string source, ref int tail, out string prerelease)
+        {
+            if (tail < source.Length && source[tail] == '-')
+            {
+                int end = source.IndexOf('+', tail);
+
+                if (end < 0)
+                {
+                    end = source.Length;
+                }
+
+                if (tail == end - 1)
+                {
+                    prerelease = null;
+                    return false;
+                }
+
+                if (!IsValidMetadataSection(source, tail + 1, end))
+                {
+                    prerelease = null;
+                    return false;
+                }
+
+                prerelease = source.Substring(tail + 1, end - tail - 1);
+
+                if (!ValidateAllDigitSegments(prerelease))
+                {
+                    prerelease = null;
+                    return false;
+                }
+
+                tail = end;
+                return true;
+            }
+
+            prerelease = null;
+            return true;
+        }
+
+        private static bool ValidateAllDigitSegments(string prerelease)
+        {
+            string[] parts = prerelease.Split('.');
+
+            for (int i = 0; i < parts.Length; ++i)
+            {
+                if (!IsNumericSegment(parts[i], out bool isInvalidBecauseOfLeadingZero) && isInvalidBecauseOfLeadingZero)
+                {
```

Comment:
The name of this method makes me think it's checking tat all segments are numeric. But the implementation appears to return true unless there is a segment that starts with a lead zero but is otherwise numeric.